src/session/content/message_row/mod.rs

```diff
@@ -228,119 +119,137 @@ pub(crate) trait MessageRowExt: IsA<MessageRow> {
+                    (
+                        match content {
+                            MessageContent::MessagePhoto(_) => {
+                                self.get_or_create_content(MessagePhoto::new).upcast()
+                            }
+                            MessageContent::MessageSticker(_) => {
+                                self.get_or_create_content(MessageSticker::new).upcast()
+                            }
+                            _ => self.get_or_create_content(MessageText::new).upcast(),
+                        },
+                        message_.is_outgoing(),
+                    )
+                } else {
+                    (self.get_or_create_content(MessageText::new).upcast(), false)
+                };
+
+            if is_outgoing {
                 content.add_css_class("outgoing");
             } else {
-                content.set_margin_start(0);
-                content.set_margin_end(AVATAR_SIZE + SPACING);
                 content.remove_css_class("outgoing");
             }
-        }
 
-        imp.message.replace(message);
-        self.notify("message");
-    }
-
-    fn connect_message_notify<F: Fn(&Self, &glib::ParamSpec) + 'static>(
-        &self,
-        f: F,
-    ) -> glib::SignalHandlerId {
-        self.connect_notify_local(Some("message"), f)
-    }
-
-    fn content(&self) -> Option<gtk::Widget> {
-        self.upcast_ref().imp().content.borrow().to_owned()
-    }
-
-    fn set_content(&self, content: Option<gtk::Widget>) {
-        if self.content() == content {
-            return;
+            content.update_content(message);
         }
 
-        let imp = self.upcast_ref().imp();
-
-        if let Some(content) = imp.content.borrow().as_ref() {
-            content.unparent();
-        }
+        fn get_or_create_content<T, F>(&self, f: F) -> T
+        where
+            T: IsA<gtk::Widget> + IsA<MessageBase>,
+            F: Fn() -> T,
+        {
+            let mut content = self.imp().content.borrow_mut();
+            match content.as_ref().and_then(|c| c.downcast_ref::<T>()) {
+                Some(content) => content.clone(),
+                None => {
+                    if let Some(old_content) = content.as_ref() {
+                        old_content.unparent();
+                    }
 
-        if let Some(ref content) = content {
-            content.set_parent(self.upcast_ref());
+                    let content_ = f();
+                    content_.set_hexpand(true);
+
+                    content_.connect_pressed(|content, x, y| {
+                        let menu = gtk::Builder::from_resource(
+                            "/com/github/melix99/telegrand/ui/message-menu.ui",
+                        )
+                        .object::<gtk::PopoverMenu>("menu")
+                        .unwrap();
+                        menu.set_parent(content);
+                        menu.set_pointing_to(Some(&gdk::Rectangle::new(x as i32, y as i32, 0, 0)));
```

Comment:
No idea why, but when the content is a MessageText, the popover menu doesn't point precisely to the cursor position.

Reply:
Huh, it's the padding. Including the padding inside the main content widget makes the actual allocation smaller.